private/buf/buflsp/buflsp.go

```diff
@@ -0,0 +1,212 @@
+//
+// Returns a context for managing the server.
+func Serve(
+	ctx context.Context,
+	container appext.Container,
+	stream jsonrpc2.Stream,
+) (jsonrpc2.Conn, error) {
+	controller, err := bufcli.NewController(container)
+	if err != nil {
+		return nil, err
+	}
+
+	bucketProvider := storageos.NewProvider(storageos.ProviderWithSymlinks())
+	bucket, err := bucketProvider.NewReadWriteBucket("/",
+		storageos.ReadWriteBucketWithSymlinksIfSupported())
+	if err != nil {
+		return nil, err
+	}
+
+	conn := jsonrpc2.NewConn(stream)
+	server := &server{
+		conn:   conn,
+		client: protocol.ClientDispatcher(conn, container.Logger()),
+
+		logger:     container.Logger(),
+		tracer:     tracing.NewTracer(container.Tracer()),
+		controller: controller,
+		rootBucket: bucket,
+	}
+	server.files = newFiles(server)
+	off := protocol.TraceOff
+	server.traceValue.Store(&off)
+
+	conn.Go(ctx, server.makeHandler())
+	return conn, nil
+}
+
+// server is an LSP server.
+//
+// This type contains all the necessary book-keeping for keeping the server running.
+// Its handler methods are not defined in buflsp.go; they are defined in other files, grouped
+// according to the groupings in https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification
+type server struct {
+	nyiServer
+
+	conn   jsonrpc2.Conn
+	client protocol.Client
+
+	logger     *zap.Logger
+	tracer     tracing.Tracer
+	controller bufctl.Controller
+	rootBucket storage.ReadBucket
+	files      *files
+
+	wktModuleSet bufmodule.ModuleSet
+
+	// These are atomics, because they are read often add written to
+	// almost never, but potentially concurrently. Having them side-by-side
+	// is fine; they are almost never written to so false sharing is not a
+	// concern.
+	initParams atomic.Pointer[protocol.InitializeParams]
+	traceValue atomic.Pointer[protocol.TraceValue]
+}
+
+// checkInit is a helper that checks if initialization has occurred and, if not,
+// returns an appropriate error.
+func (server *server) checkInit() error {
+	if server.initParams.Load() != nil {
+		return nil
+	}
+	return ErrNotInit
+}
+
+// init performs *actual* initialization of the server. This is called by Initialize().
+//
+// It may only be called once for a given server.
+func (server *server) init(ctx context.Context, params *protocol.InitializeParams) error {
+	if server.initParams.Load() != nil {
+		return fmt.Errorf("called the %q method more than once", protocol.MethodInitialize)
+	}
+	server.initParams.Store(params)
+
+	// TODO: set up logging. We need to forward everything from server.logger through to
+	// the client, if tracing is turned on. The right way to do this is with an extra
+	// goroutine and some channels.
+
+	// Load the WKTs asap. They're always needed and don't need to hit the filesystem.
+	if err := server.loadWTKs(ctx); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// loadWKTs loads a ModuleSet for the well-known types.
```

Comment:
You can probably get away with not compiling the module for WKTs here, I think... you can probably traverse the in-memory `ReadBucket` here directly and call `FindOrCreate` on each file and add them to the server files and go from there...